ampd/src/stacks/verifier.rs

```diff
@@ -0,0 +1,649 @@
+
+        if contract_log.topic != PRINT_TOPIC {
+            return Ok(false);
+        }
+
+        let tuple_type_signature = TupleTypeSignature::try_from(vec![
+            (
+                ClarityName::from("type"),
+                TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::ASCII(
+                    BufferLength::try_from(15u32)?,
+                ))),
+            ),
+            (
+                ClarityName::from("signers-hash"),
+                TypeSignature::SequenceType(SequenceSubtype::BufferType(BufferLength::try_from(
+                    32u32,
+                )?)),
+            ),
+        ])?;
+
+        let hex = contract_log
+            .value
+            .hex
+            .strip_prefix("0x")
+            .ok_or(Error::PropertyEmpty)?;
+
+        let value =
+            Value::try_deserialize_hex(hex, &TypeSignature::TupleType(tuple_type_signature), true)?;
+
+        if let Value::Tuple(data) = value {
+            if !data.get("type")?.eq(&Value::string_ascii_from_bytes(
+                SIGNERS_ROTATED_TYPE.as_bytes().to_vec(),
+            )?) {
+                return Ok(false);
+            }
+
+            let weighted_signers = WeightedSigners::try_from(&self.verifier_set)?;
+
+            let hash = weighted_signers.hash();
+
+            if !data
+                .get("signers-hash")?
+                .eq(&Value::buff_from(hash?.to_vec())?)
+            {
+                return Ok(false);
+            }
+
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a Transaction,
+    gateway_address: &String,
+    log_index: u64,
+) -> Option<&'a TransactionEvents> {
+    let event = transaction
+        .events
+        .iter()
+        .find(|el| el.event_index == log_index)?;
+
+    if !event.contract_log.as_ref()?.contract_id.eq(gateway_address) {
+        return None;
+    }
+
+    Some(event)
+}
+
+pub async fn verify_message(
+    source_chain: &ChainName,
+    gateway_address: &String,
+    transaction: &Transaction,
+    message: &Message,
+    http_client: &Client,
+) -> Vote {
+    if message.message_id.tx_hash != transaction.tx_id.as_bytes() {
+        return Vote::NotFound;
+    }
+
+    match find_event(transaction, gateway_address, message.message_id.event_index) {
+        Some(event) => {
+            if !message.eq_event(event, None).unwrap_or(false) {
+                return Vote::NotFound;
+            }
+
+            if &message.destination_chain != source_chain
+                && message.destination_address != message.source_address
+            {
+                return Vote::SucceededOnChain;
+            }
+
+            // In case message is from Stacks -> Stacks and the same contract to itself,
+            // try to see if we need to verify if a contract was deployed
+            let result = get_verify_contract_params(event);
```

Comment:
I still don't think this functionality should exist in ampd. Suddenly now ampd is doing something different than what it normally does (verify gmp messages and verify verifier set rotations). This change opens the door for ampd to start doing all kinds of custom logic, and just makes the code and system more complicated to reason about and maintain. Even from a performance perspective, we need to be aware of the overhead of attempting to verify contract deployments now. I don't think we should go down this route.

Reply:
The contract verification logic was removed, no custom logic for Stacks is now in ampd.